scvi/model/base/_archesmixin.py

```diff
@@ -155,6 +154,78 @@ def load_query_data(
 
         return model
 
+    @staticmethod
+    def prepare_query_anndata(
+        adata: AnnData,
+        reference_model: Union[str, BaseModelClass],
+        return_reference_var_names: bool = False,
+        inplace: bool = True,
+    ) -> Optional[Union[AnnData, pd.Index]]:
+        """
+        Prepare data for query integration.
+
+        This function will return a new AnnData object with padded zeros
+        for missing features, as well as correctly sorted features.
+
+        Parameters
+        ----------
+        adata
+            AnnData organized in the same way as data used to train model.
+            It is not necessary to run setup_anndata,
+            as AnnData is validated against the saved `scvi` setup dictionary.
+        reference_model
+            Either an already instantiated model of the same class, or a path to
+            saved outputs for reference model.
+        return_reference_var_names
+            Only load and return reference var names if True.
+        inplace
+            Whether to subset and rearrange query vars inplace or return new AnnData.
+
+        Returns
+        -------
+        Query adata ready to use in `load_query_data` unless `return_reference_var_names`
+        in which case a pd.Index of reference var names is returned.
+        """
+        _, var_names, _ = _get_loaded_data(reference_model)
+        var_names = pd.Index(var_names)
+
+        if return_reference_var_names:
+            return var_names
+
+        intersection = adata.var_names.intersection(var_names)
+        inter_len = len(intersection)
+        if inter_len == 0:
+            msg = "No reference var names found in query data. "
+            msg += "Please rerun with return_reference_var_names=True "
+            msg += "to see reference var names."
+            raise ValueError(msg)
+
+        ratio = inter_len / len(var_names)
+        logger.info("Found {}% reference vars in query data.".format(ratio * 100))
+        if ratio < 0.8:
+            warnings.warn(
+                "Query data contains less than 80% of reference var names. "
+                "This may result in poor performance."
+            )
+        genes_to_add = var_names.difference(adata.var_names)
+        adata_padding = AnnData(csr_matrix(np.zeros((adata.n_obs, len(genes_to_add)))))
+        adata_padding.var_names = genes_to_add
+        adata_padding.obs_names = adata.obs_names
+        # Concatenate object
+        adata_out = anndata.concat(
+            [adata, adata_padding],
+            axis=1,
+            join="outer",
+            index_unique=None,
+            merge="unique",
+        )
+        adata_out._inplace_subset_var(var_names)
+
+        if inplace:
+            adata._init_as_actual(adata_out, dtype=adata._X.dtype)
```

Comment:
if we start using this a lot, we might want to wrap it in a util since its not guaranteed stable

Reply:
agree, though now as part of scverse we can do what we want! right?!

Reply:
🤔